File: app/ocr.py
import os
from google import genai
from google.genai import types
from pydantic import BaseModel, Field
from typing import List, Optional

class InvoiceItem(BaseModel):
    product_name: str
    quantity: float
    rate: float
    amount: float

class InvoiceExtraction(BaseModel):
    invoice_date: Optional[str] = Field(None, description="Invoice date in YYYY-MM-DD format")
    items: List[InvoiceItem]

def extract_invoice_items(image_bytes: bytes, mime_type: str) -> InvoiceExtraction:
    """
    Calls Gemini model using the google-genai SDK to extract items from an invoice image.
    Uses structured outputs to guarantee JSON matching InvoiceExtraction schema.
    """
    api_key = os.getenv("GEMINI_API_KEY")
    if not api_key:
        raise ValueError("GEMINI_API_KEY is not set in the environment variables")
        
    client = genai.Client(api_key=api_key)
    
    prompt = """
    Extract ONLY inventory items from the invoice image.
    Ignore:
    - GST
    - Address
    - Vendor details
    - Phone numbers
    - Invoice totals

    For every item, extract:
    - product_name
    - quantity
    - rate
    - amount

    Also extract the invoice date and format it as YYYY-MM-DD.
    """
    
    response = client.models.generate_content(
        model='gemini-2.5-flash',
        contents=[
            types.Part.from_bytes(
                data=image_bytes,
                mime_type=mime_type,
            ),
            prompt
        ],
        config=types.GenerateContentConfig(
            response_mime_type="application/json",
            response_schema=InvoiceExtraction,
            temperature=0.1,
        ),
    )
    
    # Parse the response text using Pydantic validation
    return InvoiceExtraction.model_validate_json(response.text)

File: web/src/lib/ocr.ts
import { GoogleGenAI } from '@google/genai';

const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

const PROMPT = `You are an invoice parsing engine.
Extract inventory line items only.

Ignore:
GST
Tax totals
Addresses
Phone numbers
Bank details
Invoice summaries

Return JSON:
{
  "invoice_date":"",
  "items":[
    {
      "product_name":"",
      "quantity":0,
      "rate":0,
      "amount":0,
      "hsn":"",
      "discount":0,
      "tax_rate":0,
      "confidence_score":0.99
    }
  ]
}

If uncertain, preserve original product name exactly.
Do not invent values.
Return valid JSON only.`;

export interface ExtractedItem {
  product_name: string;
  quantity: number;
  rate: number;
  amount: number;
  hsn?: string;
  discount?: number;
  tax_rate?: number;
  confidence_score?: number;
}

export interface ExtractedInvoice {
  invoice_date: string;
  items: ExtractedItem[];
}

export async function extractInvoiceItems(
  base64Image: string, 
  mimeType: string, 
  existingProducts: string[] = []
): Promise<ExtractedInvoice> {
  let finalPrompt = PROMPT;
  if (existingProducts.length > 0) {
    finalPrompt += `\n\nCRITICAL INSTRUCTION FOR PRODUCT MAPPING:
We have a list of existing product names in our inventory database. For each extracted invoice line item, look at its name/description and check if it is a semantic or close match to any of the product names in the provided list.
- If it matches one of our existing product names (even with minor spelling differences, different case, abbreviations, extra description words, or different party/brand naming styles), return the EXACT name from our existing products list in the "product_name" field.
- If it is a completely new product that is not present in our list, return the original product name as it appears on the invoice in the "product_name" field.

List of existing products in our database:
${existingProducts.map(p => `- ${p}`).join('\n')}`;
  }

  const response = await ai.models.generateContent({
    model: 'gemini-2.5-flash',
    contents: [
      {
        role: 'user',
        parts: [
          { text: finalPrompt },
          { inlineData: { data: base64Image, mimeType } }
        ]
      }
    ],
    config: {
      responseMimeType: "application/json",
    }
  });

  if (!response.text) {
    throw new Error("Failed to extract data from image");
  }

  try {
    return JSON.parse(response.text) as ExtractedInvoice;
  } catch (e) {
    throw new Error("Invalid JSON response from AI");
  }
}
